### PR TITLE
Add EVA voice biometrics

### PR DESCRIPTION
**Pull request by jamielister**
New rule for EVA voice biometrics. Pull Request required for inclusion into Auth0 marketplace.

### CODE REVIEW

File: src/rules/eva-voice-biometric.js
Comment (joshcanhelp) [inline src/rules/eva-voice-biometric.js:99]: I wonder if this value shouldn't be `"enroll"` instead? Unless your customers are primarily Australian, this spelling might be unexpected. Your call 👍

File: src/rules/eva-voice-biometric.js
Comment (joshcanhelp) [inline src/rules/eva-voice-biometric.js:31]: Can we add a required configuration gate like the one here?

https://github.com/auth0/rules/blob/master/src/rules/scaled-access-relationships-claim.js#L22-L31

File: src/rules/eva-voice-biometric.js
Comment (joshcanhelp) [inline src/rules/eva-voice-biometric.js:36]:
```suggestion
	const clientSecret = configuration.CLIENT_SECRET;
```

File: src/rules/eva-voice-biometric.js
Comment (joshcanhelp) [inline src/rules/eva-voice-biometric.js:46]: The first 3 parameter values here are in the parent function's scope so no need to pass them in.

File: src/rules/eva-voice-biometric.js
Comment (joshcanhelp) [inline src/rules/eva-voice-biometric.js:102]:
```suggestion
		let personalDigits = ''; 
```

File: src/rules/eva-voice-biometric.js
Comment (joshcanhelp) [inline src/rules/eva-voice-biometric.js:120]:
```suggestion
						hostname: context.request.hostname,
```

To account for custom auth domains

**Review by joshcanhelp — CHANGES_REQUESTED** [review]
A few pieces of feedback and suggestions here. Also, can you run `npm run format` to normalize whitespace?

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:40]: sure, changing to auraya_eva

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:7]: I put the installation guide text in the slack channel as Markdown. I can commit it here too - is that the process?

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:33]: done

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:30]: done

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:60]: I want the jti to exist until "redirect-callback", not persisted beyond that - since it is just there to validate the temporary token. If this property is lost when the rule reenters, then I'm not sure why it is still working. Would 'context' be a better object to store these kinds of temporary properties?

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:91]: ok

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:45]: ok

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:96]: its what we use in all our docs and code. "enroll" is only standard in US - all other places it is the alternative.

File: src/rules/eva-voice-biometric.js
Comment (joshcanhelp) [inline src/rules/eva-voice-biometric.js:44]: Are these actually required, as in the Rule will not run properly without them, or will EVA provide a default?

File: src/rules/eva-voice-biometric.js
Comment (joshcanhelp) [reply src/rules/eva-voice-biometric.js]: No, sorry if my comment was unclear. We want to reduce duplication so all this description should be in the listing and we'll just link there from here (when the listing is live). This Rule needs to be merged first, then we'll correct later.

File: src/rules/eva-voice-biometric.js
Comment (joshcanhelp) [reply src/rules/eva-voice-biometric.js]: Understood. Just wanted to make sure it was intensional in a machine-name context 👍

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:44]: I'll make them optional with sensible defaults.

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js]: i've removed the description.

**Review by joshcanhelp — COMMENTED** [review]
I left a few more comments, nothing major. I'll send this to our Security team for final review before we merge it in. 

Thanks @jamielister!

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:103]: `enrolment_succeeded` means the user completed an enrolment process (this happens for the first time user). The metadata should then reflect that. Future interactions are then verifications.
`verification_accepted` means there was a verification process and it worked.

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:63]: I added a comment about this `jti` value: my understanding is that unlike `state`, this value can't be spoofed by DNS hacking or inspecting the payload since you need the secret key to decode it.

File: src/rules/eva-voice-biometric.js
Comment (joshcanhelp) [inline src/rules/eva-voice-biometric.js:45]: This is expiration just for the redirect out? In other words, if the voice check takes 3 minutes. this won't fail completely?

**Comment by jamielister** [comment]
Correct, it just for the outbound redirect and to account for clock skew. A new JWT is created by EVA for the return callback.



Regards,


Jamie Lister
CTO

Tel: +61 414 948 438
Email: jamie.lister@aurayasystems.com
Web: aurayasystems.com <https://aurayasystems.com/>

 <https://aurayasystems.com/>	
 <https://twitter.com/aurayasystems>	
 <https://www.linkedin.com/in/listerjamie/>
This email and any attachments to it are confidential and is intended for the specified recipient(s) only. It is strictly forbidden to use, share or act on the email if you are not the intended recipient(s) or if you do not have written consent from the sender. If you have received this email by mistake, please inform us by contacting the sender and deleting the original email. Please do not print this email unless it is necessary.

File: src/rules/eva-voice-biometric.js
Comment (jamielister) [reply src/rules/eva-voice-biometric.js:45]: Correct, it just for the outbound redirect and to account for clock skew. A new JWT is created by EVA for the return callback.

Comment (jamielister) [reply src/rules/eva-voice-biometric.js:45]: However, there is a (configurable) timeout for each audio prompt of about 1 minute within the EVA widget. This is to guard against hacking attempts.

**Review by joshcanhelp — APPROVED** [review]

**Pull request merged**
